lib/assent/strategies/apple.ex

```diff
@@ -0,0 +1,99 @@
+defmodule Assent.Strategy.Apple do
+  @moduledoc """
+  Apple Sign In OAuth 2.0 strategy.
+
+  You'll need to collect the 10-char long Team ID, the Services ID, the 10-char
+  Key ID and download the private key from the portal. Save the private key to
+  an accessible folder, or alternatively set `:private_key` with the content of
+  the private key.
+
+  ## Usage
+
+      config = [
+        client_id: "REPLACE_WITH_SERVICES_ID",
+        team_id: "REPLACE_WITH_TEAM_ID",
+        private_key_id: "REPLACE_WITH_PRIVATE_KEY_ID",
+        private_key_path: "/path/to/key.p8"
+      ]
+
+  ## With JS SDK
+
+  You can use the JS SDK instead of handling it through `auhorize_url/2`. All you
+  have to do is to set up you HTML page with the following way:
+
+      <html>
+        <head>
+            <meta name="appleid-signin-client-id" content="[CLIENT_ID]">
+            <meta name="appleid-signin-scope" content="[SCOPES]">
+            <meta name="appleid-signin-redirect-uri" content="[REDIRECT_URI]">
+            <meta name="appleid-signin-state" content="[STATE]">
+        </head>
+        <body>
+            <div id="appleid-signin" data-color="black" data-border="true" data-type="sign in"></div>
+            <script type="text/javascript" src="https://appleid.cdn-apple.com/appleauth/static/jsapi/appleid/1/en_US/appleid.auth.js"></script>
+        </body>
+      </html>
+
+  You can generate a state by calling:
+
+      {:ok, %{session_params: %{state: state}}} = Assent.Strategy.Apple.authorize_url(config)
+  """
+  use Assent.Strategy.OAuth2.Base
+
+  alias Assent.{Config, Strategy.OAuth2, Strategy.OAuth2.Base}
+
+  @spec default_config(Config.t()) :: Config.t()
+  def default_config(_config) do
+    [
+      site: "https://appleid.apple.com",
+      authorize_url: "/auth/authorize",
+      token_url: "/auth/token",
+      authorization_params: [scope: "email", response_mode: "form_post"],
+      auth_method: :client_secret_post
+    ]
+  end
+
+  @spec callback(Keyword.t(), map()) :: {:ok, %{user: map()}} | {:error, term()}
+  def callback(config, params) do
+    with {:ok, client_secret} <- gen_client_secret(config) do
+      config
+      |> Config.put(:client_secret, client_secret)
+      |> Base.callback(params, __MODULE__)
+    end
+  end
+
+  def gen_client_secret(config) do
+    config =
+      config
+      |> default_config()
+      |> Keyword.merge(config)
+
+    with {:ok, site}            <- Config.fetch(config, :site),
+         {:ok, client_id}       <- Config.fetch(config, :client_id),
+         {:ok, team_id}         <- Config.fetch(config, :team_id),
+         {:ok, _private_key_id} <- Config.fetch(config, :private_key_id),
+         {:ok, private_key}     <- OAuth2.__load_private_key__(config) do
+
+      timestamp = DateTime.to_unix(DateTime.utc_now())
+      claims  = %{
+        "aud" => site,
+        "iss" => team_id,
+        "sub" => client_id,
+        "iat" => timestamp,
+        "exp" => timestamp + 600
```

Comment:
Can we move this into a module attr and give it a name that represents the intention of the value?

Reply:
I'm not sure if that's needed, I'm actually think about lowering the expiration. This is only meant for communicating with Apple Sign In, and it's intended to only be consumed once to fetch an access token.

Reply:
Right, could you give it a name that brings more clarity on the code? I don't know exactly why you adding 600 (or any other value) nor you can't document things based on a name from a module attr

For example:

```
@my_random_number_exp
```

If you mention that in your code, at least I could search for it.

So the idea is to make it more clear, for us.